gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManagementStateStore.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gobblin.service.modules.orchestration;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import com.google.common.collect.Lists;
+
+import lombok.Getter;
+import lombok.Synchronized;
+
+import org.apache.gobblin.annotation.Alpha;
+import org.apache.gobblin.runtime.api.DagActionStore;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+
+
+/**
+ * An implementation of {@link DagStateStore} to provide information about dags, dag nodes and their job states.
+ * Currently, this store maintains and utilizes in-memory references about dags and their job states and is used
+ * to determine what the current status of the {@link Dag} and/or {@link Dag.DagNode} is and what actions needs to be
+ * taken next likewise mark it as: complete, failed, sla breached or simply clean up after completion.
+ * Going forward, each of these in-memory references will be read/write from MySQL store.
+ * Thus, the {@link DagManager} would then be stateless and operate independently.
+ */
+@Getter(onMethod_={@Synchronized})
+@Alpha
+public class DagManagementStateStore {
```

Comment:
the javadoc called this an impl.--and I agree it should have an interface for clients to write to!-- but I don't see that here.  as soon as we have a db-backed impl, we'll want to replace this in-memory version, but w/o updating client code using it.

Reply:
Adding to this I suggest we create an interface with additional functions for adding/checking/deleting dags in each particular status e.g. `failedDagIds.contains(), failedDagIds.add()`, etc. This will make the db-backed impl easier to write/reason about while also ideally preventing the edge cases where we make the DB implementation too similar to the in-memory implementation of returning a list of dags that match a status - which can cause a lot of GC issues in the scenario any of the lists/maps grow large in size.

Reply:
both the above suggestions makes sense and I have updated the code in my latest commit to incorporate your suggestions.